### ISSUE TITLE
multiple memory data layer as input

### ISSUE THREAD

**Issue by tachiang** [BVLC/caffe#4382]
I use the net below with multiple data inputs in training phase, and it generates corresponding weights file. Then, I try to use the generated weights file and net in my cplusplus program to predict new data in memory accessing mode.
 
# TRAINING NET:
name: "XXXNet"
layer {
  name: "data_dfg"
  type: "Data"
  top: "data_dfg"
  top: "label"
  include {
    phase: TRAIN
  }
  transform_param {
    scale: 1.0 
  }
  data_param {
    source: "examples/spatialmapping/dfg_train_lmdb"
    batch_size: 100
    backend: LMDB
  }
}
layer {
  name: "data_pea"
  type: "Data"
  top: "data_pea"
  include {
    phase: TEST
  }
  transform_param {
    scale: 0.0625 
  }
  data_param {
    source: "examples/spatialmapping/pea_test_lmdb"
    batch_size: 100
    backend: LMDB
  }
}
layer {
  name: "ip1_dfg"
  type: "InnerProduct"
  bottom: "data_dfg"
  top: "ip1_dfg"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 100  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "ip1_pea"
  type: "InnerProduct"
  bottom: "data_pea"
  top: "ip1_pea"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 100  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
	name: "concat3"
	bottom: "ip1_dfg"
	bottom: "ip1_pea"
 	top: "concat3"
	type: "Concat"
	concat_param {
		concat_dim: 1
	}
}
layer {
  name: "ip3"
  type: "InnerProduct"
  bottom: "concat3"
  top: "ip3"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 2  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "ip3"
  bottom: "label"
  top: "loss"
}

The memory accessing mode net is modified as bellow:

# PREDICTION NET:
name: "XXXNet"
layer {
  name: "data_dfg"
  type: "MemoryData"
  top: "data_dfg"
  top: "label"
  include {
    phase: TEST
  }
  transform_param {
    scale: 1.0 
  }
  memory_data_param {
    batch_size: 1
    channels: 1
    height: 10
    width: 10
  }
}
layer {
  name: "data_pea"
  type: "MemoryData"
  top: "data_pea"
  include {
    phase: TEST
  }
  transform_param {
    scale: 0.0625 
  }
  memory_data_param {
    batch_size: 1
    channels: 1
    height: 4
    width: 4
  }
}
layer {
  name: "ip1_dfg"
  type: "InnerProduct"
  bottom: "data_dfg"
  top: "ip1_dfg"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 100  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "ip1_pea"
  type: "InnerProduct"
  bottom: "data_pea"
  top: "ip1_pea"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 100  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
	name: "concat3"
	bottom: "ip1_dfg"
	bottom: "ip1_pea"
 	top: "concat3"
	type: "Concat"
	concat_param {
  		concat_dim: 1
	  }
}
layer {
  name: "ip3"
  type: "InnerProduct"
  bottom: "concat3"
  top: "ip3"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 2  
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "ip3"
  bottom: "label"
  top: "loss"
}

when I use this prediction net in cpp file, it generates errors as below:

I0629 13:21:02.820920  3950 layer_factory.hpp:77] Creating layer data_dfg
I0629 13:21:02.821046  3950 net.cpp:106] Creating Layer data_dfg
I0629 13:21:02.821132  3950 net.cpp:411] data_dfg -> data_dfg
I0629 13:21:02.821316  3950 net.cpp:411] data_dfg -> label
I0629 13:21:02.821498  3950 net.cpp:150] Setting up data_dfg
I0629 13:21:02.821574  3950 net.cpp:157] Top shape: 1 1 10 10 (100)
I0629 13:21:02.821625  3950 net.cpp:157] Top shape: 1 (1)
I0629 13:21:02.821666  3950 net.cpp:165] Memory required for data: 404
I0629 13:21:02.821724  3950 layer_factory.hpp:77] Creating layer label_data_dfg_1_split
I0629 13:21:02.821800  3950 net.cpp:106] Creating Layer label_data_dfg_1_split
I0629 13:21:02.821856  3950 net.cpp:454] label_data_dfg_1_split <- label
I0629 13:21:02.821945  3950 net.cpp:411] label_data_dfg_1_split -> label_data_dfg_1_split_0
I0629 13:21:02.822026  3950 net.cpp:411] label_data_dfg_1_split -> label_data_dfg_1_split_1
I0629 13:21:02.822106  3950 net.cpp:150] Setting up label_data_dfg_1_split
I0629 13:21:02.822155  3950 net.cpp:157] Top shape: 1 (1)
I0629 13:21:02.822206  3950 net.cpp:157] Top shape: 1 (1)
I0629 13:21:02.822247  3950 net.cpp:165] Memory required for data: 412
I0629 13:21:02.822291  3950 layer_factory.hpp:77] Creating layer data_pea
I0629 13:21:02.822365  3950 net.cpp:106] Creating Layer data_pea
I0629 13:21:02.822423  3950 net.cpp:411] data_pea -> data_pea
F0629 13:21:02.822527  3950 layer.hpp:389] Check failed: ExactNumTopBlobs() == top.size() (2 vs. 1) MemoryData Layer produces 2 top blob(s) as output.
*** Check failure stack trace: ***

The correspoding cplusplus code is as follow:

CHECK_GT(FLAGS_model.size(),0) << "Need a model definition to score";
243   CHECK_GT(FLAGS_weights.size(),0) << "Need a model weights to score";
244   LOG(INFO) << "Use CPU.";
245   Caffe::set_mode(Caffe::CPU);
246   //Initialize the  caffe net.
247   Net<float> caffe_net(FLAGS_model,caffe::TEST);
248   caffe_net.CopyTrainedLayersFrom(FLAGS_weights);
249 
250   float loss = 0.0;
251   shared_ptr<caffe::MemoryDataLayer<float> > md_layer =
252       boost::dynamic_pointer_cast <caffe::MemoryDataLayer<float> >(caffe_net.layers()[0]);

So, how can I perform multiple memory-data-layer with different shapes of blob in a net in caffe?
